File: design_dynamic_array/05_concurrent_safe/python/main.py
import threading
import time
from typing import TypeVar, Generic, List

T = TypeVar('T')

class ThreadSafeVector(Generic[T]):
    """
    05 Concurrent: Thread Safe Vector
    - Protected by a Lock (Mutex)
    """
    def __init__(self):
        self._data: List[T] = []
        self._lock = threading.Lock()

    def push_back(self, val: T) -> None:
        with self._lock:
            self._data.append(val)

    def size(self) -> int:
        with self._lock:
            return len(self._data)
            
    def get_snapshot(self) -> List[T]:
        with self._lock:
            return list(self._data) # Return copy

def worker(vec: ThreadSafeVector, id: int):
    for i in range(10):
        vec.push_back(f"Thread{id}-{i}")
        time.sleep(0.001) # Simulate work

def main():
    print("=== 05 (Concurrent Safe) Python ===")
    
    vec = ThreadSafeVector[str]()
    threads = []
    
    print("Launching 4 threads...")
    for i in range(4):
        t = threading.Thread(target=worker, args=(vec, i))
        threads.append(t)
        t.start()
        
    for t in threads:
        t.join()
        
    print(f"Final Size: {vec.size()} (Expected 40)")
    
    # Snapshot
    snap = vec.get_snapshot()
    print(f"Snapshot Sample: {snap[:3]} ...")

if __name__ == "__main__":
    main()

File: design_dynamic_array/05_concurrent_safe/main.cpp
#include "ThreadSafeVector.hpp"
#include <chrono>
#include <iostream>
#include <thread>
#include <vector>

// Simulate a worker thread
void worker(int id, concurrent::ThreadSafeVector<int> &vec) {
  for (int i = 0; i < 5; ++i) {
    // Sleep to simulate work
    std::this_thread::sleep_for(std::chrono::milliseconds(10));

    // Concurrent Write
    vec.push_back(id * 100 + i);

    // Concurrent Read
    // std::cout is NOT thread safe, so output might look messy without its own
    // mutex, but the Vector itself is safe.
  }
}

int main() {
  std::cout << "=== 05 (Concurrent Safe) ===\n";

  concurrent::ThreadSafeVector<int> sharedVec;
  std::vector<std::thread> threads;

  // Launch 4 threads
  std::cout << "Launching 4 threads writing simultaneously...\n";
  for (int i = 0; i < 4; ++i) {
    threads.emplace_back(worker, i + 1, std::ref(sharedVec));
  }

  // Wait for them
  for (auto &t : threads) {
    t.join();
  }

  std::cout << "All threads finished. Final Size: " << sharedVec.size()
            << " (Expected 20)\n";

  // Verify contents
  auto snapshot = sharedVec.get_snapshot();
  std::cout << "Content Sample: ";
  for (size_t i = 0; i < snapshot.size(); ++i) {
    if (i < 10)
      std::cout << snapshot[i] << " ";
  }
  std::cout << "...\n";

  return 0;
}
